# Module 7# 

## Total Marks: 26 marks ##

## 1. Types of dirty data and their downstream effects (N/A)##
No Jupyter Notebook exercises for this section.

## 2. Identifying Dirty Data (7.5 marks)##

### Inspecting the Data ### 
In this section we will use the pandas.describe method to undertake an inspection of our data. In this step we are trying to get a feel for your data. What does it look like and what does it represent? More importantly we need to try to understand whether we have good quality data or not. 

In this step we will use an authentic dataset: the Laurelhurst crime dataset. The Laurelhurst crime dataset contains listings of crime data reports from the Seattle Police Department: 2008 – Present (see: https://data.seattle.gov/Public-Safety/Laurelhurst-Crime/9awu-7nc4). We have already downloaded a copy of this csv file for you which is at GitHub (see: ). Please use the copy that we have provided so that you can get the same results that we do.

Note that for the following code to work you will have to make sure that the 'Laurelhurst_Crime.csv' file that you downloaded is in the same directory as your Jupyter Notebook. For example, if your Jupyter Notebook is at /projects/COIT20280/Modules/Module_7/Module_7.ipynb your downloaded file will have to also be at /projects/COIT20280/Modules/Module_7/Laurelhurst_Crime.csv


 

The code cell below simply loads the data from our csv file into a DataFrame called crime_data. We will use this DataFrame to inspect our data in the following sections.

In [1]:
#load the data in the csv file to a dataframe
import pandas as pd

crime_data = pd.read_csv('Laurelhurst_crime.csv')

Before we do any inspecting let's just do a bit of an error check to make sure that what we have in our dataframe adds up with the description given by the data providers (see: https://data.seattle.gov/Public-Safety/Laurelhurst-Crime/9awu-7nc4) and something hasn't gone wrong along the way. We can check the number of rows and columns with pandas shape method (see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) for more information. 

Call the shape method now on your dataframe to see how many rows and columns you have. 

(0.5 marks)  

(3990, 17)

In the above you should see that there are 3990 rows and 17 columns. This adds up with the description given by the data provider at the time of downloading the csv. Note that if you used a more recent version of the crime dataset the row number given by shape may vary because this dataset is updated regularly. 

So now let's get a feel for what the data looks like - a manual inspection. Use the head() method (see https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) on the crime_data dataframe to see what the first 5 rows of data look like. 

(0.5 marks)

Report Number   Offense ID  Offense Start DateTime    Offense End DateTime  \
0   2020-040665  12601509306  02/01/2020 08:00:00 PM  02/02/2020 08:00:00 AM   
1   2020-035892  12498858259  01/29/2020 03:40:00 PM  01/29/2020 04:00:00 PM   
2   2020-034200  12414526693  11/25/2019 08:00:00 AM  11/25/2019 08:20:00 AM   
3   2020-901071  12220089092  01/22/2020 08:30:00 AM  01/22/2020 08:30:00 AM   
4   2020-025006  12183510818  01/21/2020 04:19:00 PM                     NaN   

          Report DateTime Group A B Crime Against Category  \
0  02/02/2020 11:12:21 AM         A               PROPERTY   
1  01/29/2020 08:56:31 PM         A               PROPERTY   
2  01/28/2020 06:26:21 PM         A               PROPERTY   
3  01/23/2020 11:32:07 AM         A               PROPERTY   
4  01/21/2020 06:30:43 PM         A               PROPERTY   

  Offense Parent Group                   Offense Offense Code Precinct Sector  \
0  MOTOR VEHICLE THEFT       Motor Vehicle Theft          240        N      U   
1        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
2        LARCENY-THEFT       Theft From Building          23D        N      U   
3        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
4        LARCENY-THEFT       Theft From Building          23D        N      U   

  Beat       MCPP                100 Block Address   Longitude   Latitude  
0   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.288761  47.659160  
1   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.282748  47.659846  
2   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.283363  47.664383  
3   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.274393  47.664892  
4   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.284818  47.667990

Note that the head() method also provides information to us on what our columns are called which also adds up with the data description provided by the data providers. You'll also note the NaN in the last row of the Offense End DateTime. Pandas uses this sentinel value NaN (Not a Number) to indicate that there is a missing value here. We will look more at this in the next section. 

### Data Profiling ###

Now we know something about what the data looks like at a low level let's do some data profiling. We are going to Use the describe method here to do so. See https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html for more information. 

### Numerical Data ###

Use the describe method on the crime_data dataframe to return summary statistics about the numeric columns only. 
Hint: this is the default behaviour of describe 

(0.5 marks)

Offense ID    Longitude     Latitude
count  3.990000e+03  3990.000000  3990.000000
mean   9.152758e+09  -122.281306    47.662431
std    3.403801e+09     0.005477     0.004358
min    7.625486e+09  -122.290185    47.649056
25%    7.651776e+09  -122.285353    47.659367
50%    7.678226e+09  -122.282756    47.663755
75%    7.698664e+09  -122.277139    47.665790
max    2.540186e+10  -122.268773    47.668297

In the above you should get back a summary of three columns: Offence ID, Longitude and Latitude. One thing you will notice almost immediately is that the first column 'Offense ID' seems a little strange in the way it is displayed. When you used the head method previously on the dataframe you would have noticed that the Offense ID is a number that looks like 12601509306 for example. However, in the display above it will be displayed in e-notation as 2.540186e+10. The display makes it difficult to understand our summary statistics for the 'Offense ID' column. Let's fix this up before we go any further.

There several ways to fix the display so that we can see our Offense ID properly. Perhaps the most straightforward though is to change the display options to display float values as suggested in this Stackoverflow post (https://stackoverflow.com/questions/38689125/how-to-get-rid-of-pandas-converting-large-numbers-in-excel-sheet-to-exponential).

Do this now in the code cell below so we aren't using e-notation - to 2 decimal places is fine for the moment.

(0.5 marks)

Now check that you have indeed changed the display by using the describe method again on your crime_data dataframe.

(0.5 marks)

Offense ID  Longitude  Latitude
count        3990.00    3990.00   3990.00
mean   9152758068.12    -122.28     47.66
std    3403801297.41       0.01      0.00
min    7625486301.00    -122.29     47.65
25%    7651776285.00    -122.29     47.66
50%    7678226455.50    -122.28     47.66
75%    7698663674.00    -122.28     47.67
max   25401859485.00    -122.27     47.67

Ok, so now we've got rid of the annoying formatting we can try to take a better look at our data. 

### Inspecting each data column of the dataframe ###

Now it's time to see what the summary statistics reveal about each data column. Does there seem to be anything amiss here? Let's walk through it...

1. count - here we have the same count for each of the Offense ID, Longitude and Latitude. Count gives us a count of all non-null values. We know from our csv file import that there are 3990 rows. Our count values here are 3990 also which indicates we have no missing or null values in these columns. 

2. mean - mean is the average of all the values in the columns. The 'Offense ID' column is simply a unique ID numeric value so the mean reported by describe doesn't tell us a lot here. This is a different story for Longitude and Latitude though. We see our mean values are 47.66 for Latitude and -122.28 for Longitude. We need to ask ourselves whether these are sensible values. Firstly, we can find out valid values for Latitude and Longitude by a simple web search (if we don't know already). Valid values for Latitude in degress are between -90 and +90 (Southern and Northern hemispheres respectively) and valid values for Longitude are between -180 and +180 (coordinates west and east of the Meridean). Our average values fall within this range so they are at least valid. In addition we can type these Latitude and Longitude values directly into Google search (type 47.66,-122.28 into the search bar) and we are provided with a map and our exact coordinates which puts us directly in Laurelhurst - right under the Seattle Children's Hospital. So not only are the average values of the Latitude and Longitude valid they also seem sensible given this crime reporting is for the area of Laurelhurst. We aren't getting results that put us in the middle of the Pacific Ocean which may make us question the quality of our data.

3. std - standard deviation is a measure of the variance of our data. For Offense ID this makes sense - we have a range of different values that are indeed quite large numbers so our std is itself quite large. Now what about Latitude and Longitude? We see that our Longitude varies by a small amount: 0.01. This seems small - which we would expect given that it is the crime reporting for a single residential neighbourhood - but not particularly accurate or precise. Our Latitude is given as 0.00 which seems very strange and indeed incorrect. We would expect our values to be small but we expect some variation in our Latitude. So is something wrong with our data or perhaps the precision with which our data is being displayed here? Well we know from our min and max reported by describe that we certainly don't have uniform data here. For std to accurately report 0 (zero) there could be no variation in our data at all. The problem is similar to the display issue we faced with 'Offense ID'. This time however, it is because we have a small number. In our initial display the default was 2 decimal places and when we changed the display to get rid of our e-notation it did nothing to change this. Let's change this now to get it to display to 10 decimal places with pd.options.display.float_format.

(0.5 marks)

Use the describe method again on your crime_date dataframe to see if it worked.

(0.5 marks)

Offense ID       Longitude        Latitude
count        3990.0000000000 3990.0000000000 3990.0000000000
mean   9152758068.1152877808 -122.2813063353   47.6624307511
std    3403801297.4066967964    0.0054766573    0.0043580503
min    7625486301.0000000000 -122.2901853590   47.6490560000
25%    7651776285.0000000000 -122.2853526600   47.6593671100
50%    7678226455.5000000000 -122.2827559750   47.6637545800
75%    7698663674.0000000000 -122.2771385180   47.6657897600
max   25401859485.0000000000 -122.2687730430   47.6682965700

3. (cont.) so now we have adjusted the display we can see that our standard deviation is now looking a lot better. The standard deviation is small but not zero. The quality of our underlying data at this level seems fine. 

4. min & max - The minimum and maximum values of the 'Offense ID' seem sensible. Note that without having more information on what is considered a valid range for this data the best we can do is rely on our knowledge of unique ID's (see for example https://en.wikipedia.org/wiki/Unique_identifier). At this point we can see that the max and min are whole numbers ranging from 7625486301 to 25401859485 - perhaps serially generated. There is nothing terribly untoward about this and it seems to comply with our knowledge of what a unique ID might look like. Minimum values of Latitude and Longitude also seem sensible. We apply the same logic as we did in Step 2 above to arrive at this determination. Confirm this to yourself by putting the minimum and maximum values of Latitude and Longitude into google and you will get the two boundary points (extreme lower left - which places you right in the middle of Union Bay - and extreme upper right - on 55th Avenue NE just north of Windemere Park) of all your data points. Note that while we don't know the exact boundaries of the area that is being reported on we can do some work ourselves to see whether the data makes sense. We know that all of the data in the Longitude and Latitude columns are at least valid values. Additionally, in the context of the reporting it seems to make sense. All of the crimes that are reported here are contained within two extreme Latitude and Longitude points that cover the residential neighbourhood of Laurelhurst and do not seem to be in areas that extend beyond this.

5. percentiles (25%, 50%, 75%) - our percentiles simply tell us how much of the data is less than this value. So for 25% and 'Offense ID' it is stating here that 25% of the data is less than 7651776285. Likewise for Longitude 25% of the data is less than the value -122.2853526600 and for Latitude 25% of the data is less than the value 47.6593671100. It is the same for the 50% and 75%. At this stage this information simply gives us a feel for our data. It is an analysis of sorts and in terms of data cleaning there is not a lot to make us suspicious of the quality of our data. There is a reasonable spread in terms of the numbers being reported here. Again, without being an expert in the particular area of the data itself and without more powerful visualisation tools we can't know for sure - but at this early stage the data at least seems to make sense. 

### Non-numerical Data ###

So now that we have done the process for the numeric columns: 'Offense ID', 'Latitude' and 'Longitude' - let's do the same for our non-numeric columns. We want to exclude all numeric types so let's do this with the exclude parameter of the pandas describe method. See https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html to see how to exclude numeric types. 

Hint: make sure to import appropriately so you do not end up with a NameError

(0.5 marks)

Report Number  Offense Start DateTime    Offense End DateTime  \
count           3990                    3987                    2686   
unique          3817                    3730                    2572   
top      2017-140687  05/07/2018 05:44:00 PM  01/25/2015 12:00:00 PM   
freq               5                       5                       3   

               Report DateTime Group A B Crime Against Category  \
count                     3990      3990                   3990   
unique                    3808         2                      3   
top     04/22/2017 03:30:00 PM         A               PROPERTY   
freq                         5      3884                   3423   

       Offense Parent Group                   Offense Offense Code Precinct  \
count                  3990                      3990         3990     3990   
unique                   19                        32           32        2   
top           LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
freq                   1564                       877          877     3979   

       Sector  Beat       MCPP                100 Block Address  
count    3990  3990       3990                             3990  
unique      3     3          5                              370  
top         U    U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE  
freq     3974  3974       3915                              413

When we look at what describe has reported here on our non-numerical data we don't have any of the annoying issues with the display that we had with the numeric types. We can proceed with inspecting each data column of the dataframe straight away.

### Inspecting each data column of the dataframe ###

1. count - remember that count gives us a count of all non-null values in each of the columns. Here we have the same count for 'Report Number', 'Report DateTime', 'Group A B', 'Crime Against Category', 'Offense Parent Group', 'Offense', 'Offense Code', 'Precinct', 'Sector', 'Beat', 'MCPP' and '100 Block Address' that we did for our numerical columns: 'Offense ID', 'Longitude' and 'Latitude'. That is, 3990. This tells us that there are no null values in these columns, which means there are no missing values we have to contend with. The same cannot be said for 'Offense Start DateTime' or 'Offense End DateTime'. Describe tells us here that there are null values in these columns that we will have to contend with. Given this information we should ask ourselves whether there is anything here which might make us suspicious of the underlying data quality. The only issue here is the 'Offense Start DateTime' and 'Offense End DateTime' but this fact doesn't seem terribly untoward. It would seem logical that not all reports would have an EndDate. Forexample - they might be still being investigated and in terms of the 'Start DateTime' this may simply be unknown. It would be something to double check if you had access to this information. At this stage we don't and we will have to be comfortable with this judgement call and make a note of it in our data cleaning report.   

2. unique - this is a rather handy statistic for us both in terms of reporting categorical data and also numeric data (which the describe method doesn't do by default) in that it can give us an indication of duplicate data. Here we can see that for all of our non-numeric data we have repeats in the data. For example, let's look at Report Number. This has a unique value of 3817. However, we have 3990 rows. This means that there are some repeats in the values. The same goes for all of the other columns. This means that there is a potential for row duplication. If we had the Report Number giving a unique value of 3990 we would know that there is no row duplication and hence no duplication in our data. This is something to flag and investigate further on. What about the other data columns. What about Offense Start DateTime and Offense End DateTime? These not only have missing values as revealed by count but they also have repeats. Again, we don't know for sure whether this is suspicious or not without knowing more about the business context. One can think of numerous reasons why there may be Offenses that have the same start and end times. There may be peak times for crimes - crimes may genuinely occur at the same time and finish at the same time. When reporting crimes the accuracy and precision may be somewhat limited as well for whatever reason. At this stage, we could assume (and report this assumption) that these duplications are not untoward. The same goes for the Report DateTime. Group A B is a data category that works well with unique. If we had anything other than 2 here we might be suspicious of the quality of the data. While the metadata information provided by the Seattle Police Department doesn't give us a huge amount of detail we could assume that this category should only have either A or B. If there were anything greater than 2 reported here it would be worth noting and investigating further. All of the other categories here seem to make sense given the metadata information provided. Therefore, after looking at the statistics for unique all of the non-numeric columns here seem to make sense apart from the fact that we will have to check for duplicates later.  

3. top & freq- together these two reported statistics tell us more about repeating values. It tells us the value that occurs the most and how many times it occurs. Here we can see that the Report Number 2018-162496 was the most repeated value occuring 5 times in our data. The other values that are reported here for the different columns provide us with some early analytical information. They are interesting at this stage but remember that our primary purpose now is to make sure we have quality data.

#### Detailed Inspection ####
Let's investigate our potential duplication issue that we flagged in the previous step. We want to find out whether we have any duplicated data. An easy way to do this is through Pandas Dataframe.duplicated() method. See the docs (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) on this method. Put the results into another dataframe as described in this tutorial (https://thispointer.com/pandas-find-duplicate-rows-in-a-dataframe-based-on-all-or-selected-columns-using-dataframe-duplicated-in-python/) and print the results to the screen.

(0.5 marks)

Empty DataFrame
Columns: [Report Number, Offense ID, Offense Start DateTime, Offense End DateTime, Report DateTime, Group A B, Crime Against Category, Offense Parent Group, Offense, Offense Code, Precinct, Sector, Beat, MCPP, 100 Block Address, Longitude, Latitude]
Index: []

The results of the above should reveal that while we had our suspicions of duplicated data this does not in fact appear to be an issue. It returns an empty dataframe. But before we get too far ahead of ourselves let's just make sure that this method is indeed working as we think it should. 

Let's start by copying our crime_data DataFrame to another dataframe we can experiment with. Make sure that this is a deep copy using .copy() method. Stackoverflow provides a pretty good discussion of how and why this should be done here: https://stackoverflow.com/questions/27673231/why-should-i-make-a-copy-of-a-data-frame-in-pandas

Hint: this is a simple 1 line piece of code 

(0.5 marks)

Now let's select a subset of data from our crime_data and put that in a new dataframe. Call this new dataframe what you want but make it a sensible name that is descriptive of the data. Use the iloc method to select the first 5 rows of data in the crime_data dataframe. See the pandas docs for more on how to use iloc to get a subset of data (https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/03_subset_data.html)

(0.5 marks)

Let's print this subset dataframe to the screen to make sure we've got everything right and selected the first five rows of data.

(0.5 marks)

Report Number   Offense ID  Offense Start DateTime    Offense End DateTime  \
0   2020-040665  12601509306  02/01/2020 08:00:00 PM  02/02/2020 08:00:00 AM   
1   2020-035892  12498858259  01/29/2020 03:40:00 PM  01/29/2020 04:00:00 PM   
2   2020-034200  12414526693  11/25/2019 08:00:00 AM  11/25/2019 08:20:00 AM   
3   2020-901071  12220089092  01/22/2020 08:30:00 AM  01/22/2020 08:30:00 AM   
4   2020-025006  12183510818  01/21/2020 04:19:00 PM                     NaN   

          Report DateTime Group A B Crime Against Category  \
0  02/02/2020 11:12:21 AM         A               PROPERTY   
1  01/29/2020 08:56:31 PM         A               PROPERTY   
2  01/28/2020 06:26:21 PM         A               PROPERTY   
3  01/23/2020 11:32:07 AM         A               PROPERTY   
4  01/21/2020 06:30:43 PM         A               PROPERTY   

  Offense Parent Group                   Offense Offense Code Precinct Sector  \
0  MOTOR VEHICLE THEFT       Motor Vehicle Theft          240        N      U   
1        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
2        LARCENY-THEFT       Theft From Building          23D        N      U   
3        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
4        LARCENY-THEFT       Theft From Building          23D        N      U   

  Beat       MCPP                100 Block Address       Longitude  \
0   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
1   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
4   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   

       Latitude  
0 47.6591600400  
1 47.6598456900  
2 47.6643825700  
3 47.6648924700  
4 47.6679900400

Now that we have copied some data let's append that to our copied crime_data Dataframe which we set up to experiment with. Let's use pandas append method to do so. See the pandas documentation here for how to do so (see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.append.html)

Hint: Note that the result of applying append is to return a new DataFrame object. It does not change the original DataFrame to which you appended. Hence, you will need to store that new DataFrame object somewhere if you are going to use it in the future (which you will need to in the steps below). 

(0.5 marks)

Now we should check that our append has actually worked as we intended and quick print to screen should confirm this.  

(0.5 marks)

Report Number   Offense ID  Offense Start DateTime    Offense End DateTime  \
0    2020-040665  12601509306  02/01/2020 08:00:00 PM  02/02/2020 08:00:00 AM   
1    2020-035892  12498858259  01/29/2020 03:40:00 PM  01/29/2020 04:00:00 PM   
2    2020-034200  12414526693  11/25/2019 08:00:00 AM  11/25/2019 08:20:00 AM   
3    2020-901071  12220089092  01/22/2020 08:30:00 AM  01/22/2020 08:30:00 AM   
4    2020-025006  12183510818  01/21/2020 04:19:00 PM                     NaN   
..           ...          ...                     ...                     ...   
0    2020-040665  12601509306  02/01/2020 08:00:00 PM  02/02/2020 08:00:00 AM   
1    2020-035892  12498858259  01/29/2020 03:40:00 PM  01/29/2020 04:00:00 PM   
2    2020-034200  12414526693  11/25/2019 08:00:00 AM  11/25/2019 08:20:00 AM   
3    2020-901071  12220089092  01/22/2020 08:30:00 AM  01/22/2020 08:30:00 AM   
4    2020-025006  12183510818  01/21/2020 04:19:00 PM                     NaN   

           Report DateTime Group A B Crime Against Category  \
0   02/02/2020 11:12:21 AM         A               PROPERTY   
1   01/29/2020 08:56:31 PM         A               PROPERTY   
2   01/28/2020 06:26:21 PM         A               PROPERTY   
3   01/23/2020 11:32:07 AM         A               PROPERTY   
4   01/21/2020 06:30:43 PM         A               PROPERTY   
..                     ...       ...                    ...   
0   02/02/2020 11:12:21 AM         A               PROPERTY   
1   01/29/2020 08:56:31 PM         A               PROPERTY   
2   01/28/2020 06:26:21 PM         A               PROPERTY   
3   01/23/2020 11:32:07 AM         A               PROPERTY   
4   01/21/2020 06:30:43 PM         A               PROPERTY   

   Offense Parent Group                   Offense Offense Code Precinct  \
0   MOTOR VEHICLE THEFT       Motor Vehicle Theft          240        N   
1         LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
2         LARCENY-THEFT       Theft From Building          23D        N   
3         LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
4         LARCENY-THEFT       Theft From Building          23D        N   
..                  ...                       ...          ...      ...   
0   MOTOR VEHICLE THEFT       Motor Vehicle Theft          240        N   
1         LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
2         LARCENY-THEFT       Theft From Building          23D        N   
3         LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
4         LARCENY-THEFT       Theft From Building          23D        N   

   Sector Beat       MCPP                100 Block Address       Longitude  \
0       U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
1       U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2       U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3       U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
4       U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   
..    ...  ...        ...                              ...             ...   
0       U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
1       U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2       U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3       U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
4       U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   

        Latitude  
0  47.6591600400  
1  47.6598456900  
2  47.6643825700  
3  47.6648924700  
4  47.6679900400  
..           ...  
0  47.6591600400  
1  47.6598456900  
2  47.6643825700  
3  47.6648924700  
4  47.6679900400  

[3995 rows x 17 columns]

In the above you should check that you have an extra 5 rows. You should also see here that you do indeed have repeats. Printing to screen will print the first 5 rows - which were the rows we copied - and the last 5 rows - which were the copied rows that we appended. 

Now let's apply the duplicate method again on the dataframe that contains the duplicated data to see whether it works as we think it should. As you did previously put these results into a new dataframe and print this dataframe to the screen. 

(0.5 marks)

Report Number   Offense ID  Offense Start DateTime    Offense End DateTime  \
0   2020-040665  12601509306  02/01/2020 08:00:00 PM  02/02/2020 08:00:00 AM   
1   2020-035892  12498858259  01/29/2020 03:40:00 PM  01/29/2020 04:00:00 PM   
2   2020-034200  12414526693  11/25/2019 08:00:00 AM  11/25/2019 08:20:00 AM   
3   2020-901071  12220089092  01/22/2020 08:30:00 AM  01/22/2020 08:30:00 AM   
4   2020-025006  12183510818  01/21/2020 04:19:00 PM                     NaN   

          Report DateTime Group A B Crime Against Category  \
0  02/02/2020 11:12:21 AM         A               PROPERTY   
1  01/29/2020 08:56:31 PM         A               PROPERTY   
2  01/28/2020 06:26:21 PM         A               PROPERTY   
3  01/23/2020 11:32:07 AM         A               PROPERTY   
4  01/21/2020 06:30:43 PM         A               PROPERTY   

  Offense Parent Group                   Offense Offense Code Precinct Sector  \
0  MOTOR VEHICLE THEFT       Motor Vehicle Theft          240        N      U   
1        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
2        LARCENY-THEFT       Theft From Building          23D        N      U   
3        LARCENY-THEFT  Theft From Motor Vehicle          23F        N      U   
4        LARCENY-THEFT       Theft From Building          23D        N      U   

  Beat       MCPP                100 Block Address       Longitude  \
0   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
1   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
4   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   

       Latitude  
0 47.6591600400  
1 47.6598456900  
2 47.6643825700  
3 47.6648924700  
4 47.6679900400

In the above we should see those same 5 data rows that we copied earlier. If you don't have this retrace your steps and make sure you have done everything right (particularly by referring to the docs).

#### Summary ####
In summary this section 2 has walked you through the inspection step of Elgabry's (2019) workflow for data cleaning (see https://towardsdatascience.com/the-ultimate-guide-to-data-cleaning-3969843991d4) on an authentic dataset. This was the Laurelhurst crime data set - an open data set provided by the Seattle Police Department. The next section will go into more detail on the cleaning methods available to us by continuing to follow through with this same example. This time however, the dataset has been changed to reflect a dirty dataset.  

## 3. Methods for Cleaning (9 marks)##

In this section we are going to look in more detail at the three different classes of dirty data and the methods available for cleaning them. We will be using the methods as presented by McKinney (2017) for doing so. Additionally, this time we will be using the unclean version of the Laurelhurst_crime.csv dataset: Laurelhurst_crime_unclean.csv. This csv file is also available on Github (see: ) and has been intentionally modified from the clean version to give you experience with cleaning data. Of course this seems a little around the wrong way. In practice you would be given an unclean dataset - or at least a dataset that is in a state which is unknown and so is assumed to require some level of cleaning. You would then work toward getting a clean dataset. However, for the purposes of teaching and practice going the wrong way around - seeing a clean data set and then seeing a dirty data set - will serve us well enough.        

As in Section 2 make sure that the 'Laurelhurst_crime_unclean.csv' file that you downloaded is in the same directory as your Jupyter Notebook. 

The code cell below simply loads the data from our csv file into a DataFrame called crime_data_unclean. We will use this DataFrame to inspect our data in the following sections.

In [2]:
#load the data in the csv file to a dataframe
import pandas as pd


crime_data_unclean = pd.read_csv('Laurelhurst_crime_unclean.csv')
crime_data_unclean


Report Number   Offense ID Offense Start DateTime Offense End DateTime  \
0      2020-040665  12601509306    2020-02-01 20:00:00  2020-02-02 08:00:00   
1      2020-035892  12498858259    2020-01-29 15:40:00  2020-01-29 16:00:00   
2      2020-034200  12414526693    2019-11-25 08:00:00  2019-11-25 08:20:00   
3      2020-901071  12220089092    2020-01-22 08:30:00  2020-01-22 08:30:00   
4      2020-025006  12183510818    2020-01-21 16:19:00                  NaN   
...            ...          ...                    ...                  ...   
3990   2020-040665  12601509306    2020-02-01 20:00:00  2020-02-02 08:00:00   
3991   2020-035892  12498858259    2020-01-29 15:40:00  2020-01-29 16:00:00   
3992   2020-034200  12414526693    2019-11-25 08:00:00  2019-11-25 08:20:00   
3993   2020-901071  12220089092    2020-01-22 08:30:00  2020-01-22 08:30:00   
3994   2020-025006  12183510818    2020-01-21 16:19:00                  NaN   

          Report DateTime Group A B Crime Against Category  \
0     2020-02-02 11:12:21         A               PROPERTY   
1     2020-01-29 20:56:31         A               PROPERTY   
2     2020-01-28 18:26:21         A               PROPERTY   
3     2020-01-23 11:32:07         A               PROPERTY   
4     2020-01-21 18:30:43         A               PROPERTY   
...                   ...       ...                    ...   
3990  2020-02-02 11:12:21         A               PROPERTY   
3991  2020-01-29 20:56:31         A               PROPERTY   
3992  2020-01-28 18:26:21         A               PROPERTY   
3993  2020-01-23 11:32:07         A               PROPERTY   
3994  2020-01-21 18:30:43         A               PROPERTY   

          Offense Parent Group              Offense Offense Code Precinct  \
0          Motor Vehicle Theft  MOTOR VEHICLE THEFT          240        N   
1     Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
2          Theft From Building        LARCENY-THEFT          23D        N   
3     Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
4          Theft From Building        LARCENY-THEFT          23D        N   
...                        ...                  ...          ...      ...   
3990       Motor Vehicle Theft  MOTOR VEHICLE THEFT          240        N   
3991  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3992       Theft From Building        LARCENY-THEFT          23D        N   
3993  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3994       Theft From Building        LARCENY-THEFT          23D        N   

     Sector Beat       MCPP                100 Block Address   Longitude  \
0         U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.288761   
1         U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.282748   
2         U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.283363   
3         U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.274393   
4         U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.284818   
...     ...  ...        ...                              ...         ...   
3990      U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.288761   
3991      U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.282748   
3992      U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.283363   
3993      U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.274393   
3994      U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.284818   

       Latitude  
0     47.659160  
1     47.659846  
2     47.664383  
3     47.664892  
4     47.667990  
...         ...  
3990  47.659160  
3991  47.659846  
3992  47.664383  
3993  47.664892  
3994  47.667990  

[3995 rows x 17 columns]

The above gives you a look at the data - the first five rows and the last five rows. At this stage it simply looks like our appended_crime_data dataframe that we had previously.

Let's get a quick overview of this data by applying pandas describe as we did in the previous section. This time include all columns of your data so you can see everything together.

(0.5 marks)

Report Number             Offense ID Offense Start DateTime  \
count           3995        3995.0000000000                   3992   
unique          3817                    nan                   3730   
top      2018-162496                    nan    2017-04-22 15:30:00   
freq               5                    nan                      5   
mean             NaN  9156801798.7354202271                    NaN   
std              NaN  3403592709.0212335587                    NaN   
min              NaN  7625486301.0000000000                    NaN   
25%              NaN  7651803924.5000000000                    NaN   
50%              NaN  7678342173.0000000000                    NaN   
75%              NaN  7698732698.0000000000                    NaN   
max              NaN 25401859485.0000000000                    NaN   

       Offense End DateTime      Report DateTime Group A B  \
count                  2690                 3995      3995   
unique                 2572                 3808         2   
top     2020-05-15 00:00:00  2018-05-07 20:44:00         A   
freq                      3                    5      3889   
mean                    NaN                  NaN       NaN   
std                     NaN                  NaN       NaN   
min                     NaN                  NaN       NaN   
25%                     NaN                  NaN       NaN   
50%                     NaN                  NaN       NaN   
75%                     NaN                  NaN       NaN   
max                     NaN                  NaN       NaN   

       Crime Against Category      Offense Parent Group        Offense  \
count                    3995                      3995           3995   
unique                      3                        32             19   
top                  PROPERTY  Theft From Motor Vehicle  LARCENY-THEFT   
freq                     3428                       879           1568   
mean                      NaN                       NaN            NaN   
std                       NaN                       NaN            NaN   
min                       NaN                       NaN            NaN   
25%                       NaN                       NaN            NaN   
50%                       NaN                       NaN            NaN   
75%                       NaN                       NaN            NaN   
max                       NaN                       NaN            NaN   

       Offense Code Precinct Sector  Beat       MCPP  \
count          3995     3995   3995  3995       3995   
unique           32        2      3     3          5   
top             23F        N      U    U3  SANDPOINT   
freq            879     3984   3979  3979       3920   
mean            NaN      NaN    NaN   NaN        NaN   
std             NaN      NaN    NaN   NaN        NaN   
min             NaN      NaN    NaN   NaN        NaN   
25%             NaN      NaN    NaN   NaN        NaN   
50%             NaN      NaN    NaN   NaN        NaN   
75%             NaN      NaN    NaN   NaN        NaN   
max             NaN      NaN    NaN   NaN        NaN   

                      100 Block Address       Longitude        Latitude  
count                              3995 3995.0000000000 3995.0000000000  
unique                              370             nan             nan  
top     48XX BLOCK OF SAND POINT WAY NE             nan             nan  
freq                                414             nan             nan  
mean                                NaN -122.2813082254   47.6624317817  
std                                 NaN    0.0054760174    0.0043569927  
min                                 NaN -122.2901853590   47.6490560000  
25%                                 NaN -122.2853526600   47.6593671100  
50%                                 NaN -122.2827559750   47.6637545800  
75%                                 NaN -122.2771385180   47.6657897600  
max                                 NaN -122.

### 3.1 Missing Data (2 marks) ###
When we inspected our data in the previous section we found that we had some missing data for 'Offense Start DateTime' and 'Offense End DateTime'. Pandas represents this data as NaN which is what we saw when we applied the head() method to our dataframe. 

Now there are two main methods for dealing with missing data which we will cover below: filtering out data and filling in data.


#### Filtering Out Data ####
If we were told that our analysis relied only on a complete data set - all records had to have no missing values for any of the columns - we would have to exclude these values.  In the code cell below go ahead and filter out all rows that contain missing values. 

Hint: McKinney (2017) shows us how this can be done in Section: 7.1 Handling Missing Data, Subsection: Filtering Out Missing Data

(0.5 marks)

Now apply describe to do a data profile now that you don't have records with missing values. To keep this tidy apply describe for numerical values and then describe for the other values.

(1 mark)

Offense ID       Longitude        Latitude
count        2689.0000000000 2689.0000000000 2689.0000000000
mean   9515261967.5939006805 -122.2810168858   47.6618351741
std    3677377994.2488069534    0.0058195819    0.0043502162
min    7625486301.0000000000 -122.2901853590   47.6490560000
25%    7653134941.0000000000 -122.2857977340   47.6584320200
50%    7681625641.0000000000 -122.2815428870   47.6621847000
75%    8616056676.0000000000 -122.2760584100   47.6656930800
max   25397199761.0000000000 -122.2687730430   47.6682965700

Report Number Offense Start DateTime Offense End DateTime  \
count           2689                   2689                 2689   
unique          2618                   2540                 2571   
top      2008-203723    2020-03-15 00:00:00  2020-05-15 17:00:00   
freq               3                      3                    3   

            Report DateTime Group A B Crime Against Category  \
count                  2689      2689                   2689   
unique                 2613         2                      3   
top     2008-01-29 12:00:00         A               PROPERTY   
freq                      3      2660                   2573   

            Offense Parent Group        Offense Offense Code Precinct Sector  \
count                       2689           2689         2689     2689   2689   
unique                        29             17           29        2      3   
top     Theft From Motor Vehicle  LARCENY-THEFT          23F        N      U   
freq                         781           1241          781     2683   2678   

        Beat       MCPP                100 Block Address  
count   2689       2689                             2689  
unique     3          5                              344  
top       U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE  
freq    2678       2634                              155

#### Filling In Missing Data ####
The other method we have available to us for dealing with missing data is filling in missing data.  

Hint: McKinney (2017) shows us how this can be done in Section: 7.1 Handling Missing Data, Subsection: Filling In Missing Data

We have been told that our analysis must include all available information. One method for filling in data is to use a forward fill method. This method takes the last valid observation and forward fills this value until it gets to the next valid observation. See the docs for more information (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html). 

Apply this method now on the unclean data ie. the dataframe that represents Laurelhurst_data_unclean.csv 

(0.5 marks)

Report Number   Offense ID Offense Start DateTime Offense End DateTime  \
0      2020-040665  12601509306    2020-02-01 20:00:00  2020-02-02 08:00:00   
1      2020-035892  12498858259    2020-01-29 15:40:00  2020-01-29 16:00:00   
2      2020-034200  12414526693    2019-11-25 08:00:00  2019-11-25 08:20:00   
3      2020-901071  12220089092    2020-01-22 08:30:00  2020-01-22 08:30:00   
4      2020-025006  12183510818    2020-01-21 16:19:00  2020-01-22 08:30:00   
...            ...          ...                    ...                  ...   
3990   2020-040665  12601509306    2020-02-01 20:00:00  2020-02-02 08:00:00   
3991   2020-035892  12498858259    2020-01-29 15:40:00  2020-01-29 16:00:00   
3992   2020-034200  12414526693    2019-11-25 08:00:00  2019-11-25 08:20:00   
3993   2020-901071  12220089092    2020-01-22 08:30:00  2020-01-22 08:30:00   
3994   2020-025006  12183510818    2020-01-21 16:19:00  2020-01-22 08:30:00   

          Report DateTime Group A B Crime Against Category  \
0     2020-02-02 11:12:21         A               PROPERTY   
1     2020-01-29 20:56:31         A               PROPERTY   
2     2020-01-28 18:26:21         A               PROPERTY   
3     2020-01-23 11:32:07         A               PROPERTY   
4     2020-01-21 18:30:43         A               PROPERTY   
...                   ...       ...                    ...   
3990  2020-02-02 11:12:21         A               PROPERTY   
3991  2020-01-29 20:56:31         A               PROPERTY   
3992  2020-01-28 18:26:21         A               PROPERTY   
3993  2020-01-23 11:32:07         A               PROPERTY   
3994  2020-01-21 18:30:43         A               PROPERTY   

          Offense Parent Group              Offense Offense Code Precinct  \
0          Motor Vehicle Theft  MOTOR VEHICLE THEFT          240        N   
1     Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
2          Theft From Building        LARCENY-THEFT          23D        N   
3     Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
4          Theft From Building        LARCENY-THEFT          23D        N   
...                        ...                  ...          ...      ...   
3990       Motor Vehicle Theft  MOTOR VEHICLE THEFT          240        N   
3991  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3992       Theft From Building        LARCENY-THEFT          23D        N   
3993  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3994       Theft From Building        LARCENY-THEFT          23D        N   

     Sector Beat       MCPP                100 Block Address       Longitude  \
0         U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
1         U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2         U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3         U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
4         U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   
...     ...  ...        ...                              ...             ...   
3990      U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
3991      U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
3992      U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3993      U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
3994      U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   

          Latitude  
0    47.6591600400  
1    47.6598456900  
2    47.6643825700  
3    47.6648924700  
4    47.6679900400  
...            ...  
3990 47.6591600400  
3991 47.6598456900  
3992 47.6643825700  
3993 47.6648924700  
3994 47.6679900400  

[3995 rows x 17 columns]

Let's compare what we have in the above printout with what we had when we first printed out our crime_data_unclean dataframe. We can see that the original NaN that we had appearing in the fifth and last rows for the column 'Offense End DateTime' has now been filled in. We can also see that this value is exactly the same as the value before it. 

Remember that fillna does not change the original dataframe and that if we wanted to use this dataframe going forward we should either put these results into a new dataframe or change the inplace parameter to 'True'. 

### 3.2 Surplus Data (2.5 marks) ###
Our next class of dirty data is surplus data. Surplus data includes irrelevant data (unwanted observations, unwanted outliers) and duplicate data. We already went through identifying duplicate data in our previous section. Here we will cover how to deal with unwanted observations and unwanted outliers as well as duplicate data.  


#### Irrelevant Data ####
Remember that clean and unclean data is a matter of perspective. Data can be valid and without errors but if we don't want it in our analysis for whatever reason then we can consider it as unclean in that it is not fit for purpose. 

In this section we will look at restricting our data to only certain observations. 


Let's say that we have been told we have to do an analysis on crimes that occurred in an area encompassed by the 25th and 75th percentiles of our Latitude and Longitude (based on the clean data). Your job is to create a new dataframe that is restricted to all crimes reported in what represents this area.

Hint: This post on stackoverflow provides a good walkthrough on how to select rows of a dataframe based on certain criteria of a column value (see: https://stackoverflow.com/questions/17071871/how-do-i-select-rows-from-a-dataframe-based-on-column-values). Make sure you put your selection into a new dataframe and for the purposes of this small exercise make your selection of rows from the original clean dataframe: crime_data.

(0.5 marks)

Print this to screen to make sure that you have data in your dataframe.

(0.5 marks)

Report Number   Offense ID  Offense Start DateTime  \
1      2020-035892  12498858259  01/29/2020 03:40:00 PM   
2      2020-034200  12414526693  11/25/2019 08:00:00 AM   
14     2019-917421  12117875034  12/25/2019 06:00:00 AM   
20     2019-464703  12104608721  12/14/2019 05:30:00 PM   
22     2019-462060  12084934502  12/12/2019 06:30:00 PM   
...            ...          ...                     ...   
3968   2018-434976   7663852161  11/19/2018 08:30:00 PM   
3971   2021-122036  24076015508  05/18/2021 05:46:00 PM   
3973   2021-123865  24394729109  05/16/2021 12:00:00 PM   
3976   2021-123781  24393595338  05/18/2021 03:40:00 AM   
3984   2021-134198  25354733422  05/31/2021 06:33:00 PM   

        Offense End DateTime         Report DateTime Group A B  \
1     01/29/2020 04:00:00 PM  01/29/2020 08:56:31 PM         A   
2     11/25/2019 08:20:00 AM  01/28/2020 06:26:21 PM         A   
14    12/26/2019 05:00:00 PM  12/27/2019 09:02:17 AM         A   
20    12/16/2019 11:00:00 AM  12/16/2019 03:32:18 PM         A   
22    12/13/2019 08:00:00 PM  12/14/2019 06:22:52 AM         A   
...                      ...                     ...       ...   
3968  11/20/2018 11:00:00 AM  11/20/2018 12:56:00 PM         A   
3971                     NaN  05/18/2021 08:27:38 PM         A   
3973  05/19/2021 06:00:00 PM  05/20/2021 08:09:01 PM         A   
3976                     NaN  05/20/2021 07:55:36 PM         A   
3984                     NaN  05/31/2021 08:57:24 PM         A   

     Crime Against Category        Offense Parent Group  \
1                  PROPERTY               LARCENY-THEFT   
2                  PROPERTY               LARCENY-THEFT   
14                 PROPERTY               LARCENY-THEFT   
20                 PROPERTY               LARCENY-THEFT   
22                 PROPERTY         MOTOR VEHICLE THEFT   
...                     ...                         ...   
3968               PROPERTY               LARCENY-THEFT   
3971                 PERSON            ASSAULT OFFENSES   
3973               PROPERTY         MOTOR VEHICLE THEFT   
3976               PROPERTY  BURGLARY/BREAKING&ENTERING   
3984                 PERSON            ASSAULT OFFENSES   

                                          Offense Offense Code Precinct  \
1                        Theft From Motor Vehicle          23F        N   
2                             Theft From Building          23D        N   
14                       Theft From Motor Vehicle          23F        N   
20    Theft of Motor Vehicle Parts or Accessories          23G        N   
22                            Motor Vehicle Theft          240        N   
...                                           ...          ...      ...   
3968  Theft of Motor Vehicle Parts or Accessories          23G        N   
3971                               Simple Assault          13B        N   
3973                          Motor Vehicle Theft          240        N   
3976                 Burglary/Breaking & Entering          220        N   
3984                                 Intimidation          13C        N   

     Sector Beat       MCPP                100 Block Address       Longitude  \
1         U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
2         U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
14        U   U3  SANDPOINT        47XX BLOCK OF 45TH AVE NE -122.2794048830   
20        U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
22        U   U3  SANDPOINT        42XX BLOCK OF 43RD AVE NE -122.2813890210   
...     ...  ...        ...                              ...             ...   
3968      U   U3  SANDPOINT         40XX BLOCK OF NE 50TH ST -122.2841400630   
3971      U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3973      U   U3    UNKNOWN         NE 45TH ST / 43RD AVE NE -122.2813832420   
3976      U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3984      U  

#### Duplicate Data ####
Previously we looked at identifying duplicate data. Now it's time to remove that duplicate data. Follow McKinney (2017) and remove duplicate data from the dataframe crime_data_unclean. Put this in a new dataframe named appropriately (this needn't be a deep copy).

Firstly, however, we should identify what duplicates we have in the dataframe like we did in Section 2 above. Do this in the code cell below.

(0.5 marks)

Report Number   Offense ID Offense Start DateTime Offense End DateTime  \
3990   2020-040665  12601509306    2020-02-01 20:00:00  2020-02-02 08:00:00   
3991   2020-035892  12498858259    2020-01-29 15:40:00  2020-01-29 16:00:00   
3992   2020-034200  12414526693    2019-11-25 08:00:00  2019-11-25 08:20:00   
3993   2020-901071  12220089092    2020-01-22 08:30:00  2020-01-22 08:30:00   
3994   2020-025006  12183510818    2020-01-21 16:19:00                  NaN   

          Report DateTime Group A B Crime Against Category  \
3990  2020-02-02 11:12:21         A               PROPERTY   
3991  2020-01-29 20:56:31         A               PROPERTY   
3992  2020-01-28 18:26:21         A               PROPERTY   
3993  2020-01-23 11:32:07         A               PROPERTY   
3994  2020-01-21 18:30:43         A               PROPERTY   

          Offense Parent Group              Offense Offense Code Precinct  \
3990       Motor Vehicle Theft  MOTOR VEHICLE THEFT          240        N   
3991  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3992       Theft From Building        LARCENY-THEFT          23D        N   
3993  Theft From Motor Vehicle        LARCENY-THEFT          23F        N   
3994       Theft From Building        LARCENY-THEFT          23D        N   

     Sector Beat       MCPP                100 Block Address       Longitude  \
3990      U   U3  SANDPOINT         36XX BLOCK OF NE 42ND ST -122.2887605220   
3991      U   U3  SANDPOINT        41XX BLOCK OF 42ND AVE NE -122.2827475600   
3992      U   U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE -122.2833627690   
3993      U   U3  SANDPOINT         49XX BLOCK OF NE 50TH ST -122.2743934530   
3994      U   U3  SANDPOINT        52XX BLOCK OF 40TH AVE NE -122.2848180340   

          Latitude  
3990 47.6591600400  
3991 47.6598456900  
3992 47.6643825700  
3993 47.6648924700  
3994 47.6679900400

So now we know what data has been duplicated let's go ahead and remove the duplicate data from the dataframe crime_data_unclean. Store this in a new dataframe with a suitable name.

(0.5 marks)

Now, to see that this has worked let's see if we can pick up any duplicates.

Hint: our dataframe should be empty if we print to screen

(0.5 marks)

Empty DataFrame
Columns: [Report Number, Offense ID, Offense Start DateTime, Offense End DateTime, Report DateTime, Group A B, Crime Against Category, Offense Parent Group, Offense, Offense Code, Precinct, Sector, Beat, MCPP, 100 Block Address, Longitude, Latitude]
Index: []

And now that we've removed duplicates from our dataframe we will continue to use this dataframe (the one with no duplicates but is otherwise unclean) in the remainder of our cleaning process. At the end we will write our cleaned data to file. 

### 3.3 Unrefined Data (4 marks) ###
Unrefined data includes data that contains structural errors (eg. syntax errors and type conversion errors) or is in an unstructured form. In this section we are going to go through a few of the common methods we have available to us for cleaning structural errors in data. In this section we will use the dataframe that has no duplicates but is otherwise unclean.  

#### Structural Errors ####


You may have noticed that when you applied describe() to the unclean dataframe the results were quite different to the clean dataframe. Just to make sure that you do notice the difference apply describe() again to the two different dataframes below: the clean dataframe and the dataframe with no duplicates but is otherwise unclean. 

(1 mark)

Offense ID       Longitude        Latitude
count        3990.0000000000 3990.0000000000 3990.0000000000
mean   9152758068.1152877808 -122.2813063353   47.6624307511
std    3403801297.4066967964    0.0054766573    0.0043580503
min    7625486301.0000000000 -122.2901853590   47.6490560000
25%    7651776285.0000000000 -122.2853526600   47.6593671100
50%    7678226455.5000000000 -122.2827559750   47.6637545800
75%    7698663674.0000000000 -122.2771385180   47.6657897600
max   25401859485.0000000000 -122.2687730430   47.6682965700

Report Number  Offense Start DateTime    Offense End DateTime  \
count           3990                    3987                    2686   
unique          3817                    3730                    2572   
top      2017-140687  05/07/2018 05:44:00 PM  01/25/2015 12:00:00 PM   
freq               5                       5                       3   

               Report DateTime Group A B Crime Against Category  \
count                     3990      3990                   3990   
unique                    3808         2                      3   
top     04/22/2017 03:30:00 PM         A               PROPERTY   
freq                         5      3884                   3423   

       Offense Parent Group                   Offense Offense Code Precinct  \
count                  3990                      3990         3990     3990   
unique                   19                        32           32        2   
top           LARCENY-THEFT  Theft From Motor Vehicle          23F        N   
freq                   1564                       877          877     3979   

       Sector  Beat       MCPP                100 Block Address  
count    3990  3990       3990                             3990  
unique      3     3          5                              370  
top         U    U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE  
freq     3974  3974       3915                              413

Offense ID       Longitude        Latitude
count        3995.0000000000 3995.0000000000 3995.0000000000
mean   9156801798.7354202271 -122.2813082254   47.6624317817
std    3403592709.0212335587    0.0054760174    0.0043569927
min    7625486301.0000000000 -122.2901853590   47.6490560000
25%    7651803924.5000000000 -122.2853526600   47.6593671100
50%    7678342173.0000000000 -122.2827559750   47.6637545800
75%    7698732698.0000000000 -122.2771385180   47.6657897600
max   25401859485.0000000000 -122.2687730430   47.6682965700

Report Number Offense Start DateTime Offense End DateTime  \
count           3995                   3992                 2690   
unique          3817                   3730                 2572   
top      2018-162496    2017-04-22 15:30:00  2020-05-15 00:00:00   
freq               5                      5                    3   

            Report DateTime Group A B Crime Against Category  \
count                  3995      3995                   3995   
unique                 3808         2                      3   
top     2018-05-07 20:44:00         A               PROPERTY   
freq                      5      3889                   3428   

            Offense Parent Group        Offense Offense Code Precinct Sector  \
count                       3995           3995         3995     3995   3995   
unique                        32             19           32        2      3   
top     Theft From Motor Vehicle  LARCENY-THEFT          23F        N      U   
freq                         879           1568          879     3984   3979   

        Beat       MCPP                100 Block Address  
count   3995       3995                             3995  
unique     3          5                              370  
top       U3  SANDPOINT  48XX BLOCK OF SAND POINT WAY NE  
freq    3979       3920                              414

Firstly, let's start by looking at the 'Group A B' column. You should see that there are 3 unique values for this column in the data. This doesn't make sense so let's have a closer look at what these are. In the cell below apply the .unique() method against this column of the dataframe to see what they are.

(0.5 marks)

array(['A', 'B'], dtype=object)

So now that we've applied .unique() we can see that instead of only an 'A' and a 'B' there is also an 'a'. We need to fix this somehow. The easiest way here is to simply change everything in this column to upper case. However, before we fix it let's see how many rows are effected. In the cell below select all the rows that have a value of 'a' in the 'Group A B' column. Use .loc to do so. 

(0.5 marks)

Empty DataFrame
Columns: [Report Number, Offense ID, Offense Start DateTime, Offense End DateTime, Report DateTime, Group A B, Crime Against Category, Offense Parent Group, Offense, Offense Code, Precinct, Sector, Beat, MCPP, 100 Block Address, Longitude, Latitude]
Index: []

In the above cell you should see that only one row was effected. So now let's go ahead and fix it. Let's apply the str.upper() on the 'Group A B' column. McKinney(2017) does this with the str.lower method. Use the same approach here with str.upper and don't worry about doing this in a function just yet.

(0.5 marks)

Now that you've applied the method you need to check whether it actually worked. Use .loc on the appropriate row number to find out.

(0.5 marks)

'A'

Now remember .str.upper() applied against the column will give you a series. You now need to reflect this change in the dataframe. Make the appropriate column of the dataframe ie. 'Group A B' equal the correct series. 

(0.5 marks)

Now check that it worked by seeing how many unique values for 'Group A B' are in the dataframe.

(0.5 marks)

array(['A', 'B'], dtype=object)

Your crime_data_unclean should now only have 2 unique values 'A' and 'B'.

## 4. Challenge: Find the Errors (9.5 marks) ##
In the previous two sections you were stepped through how to look for errors and what to do with them once you find them. In the following section you will be required to find a number of different errors that have been deliberately put into the unclean dataset. There are five additional errors apart from the ones we have already dealt with. 

In this section we will apply a modified version of Elgabry's (2019) workflow for data cleaning (see: https://towardsdatascience.com/the-ultimate-guide-to-data-cleaning-3969843991d4).  For each of these errors you will have to do the following:

1) Describe exactly what the error is and illustrate this appropriately with code 

2) Report how many rows and or columns are effected - produce code that confirms this

3) Show how to fix the problem commenting on your code appropriately. If you have used code from elsewhere you should reference it here. It is fine to do so but make sure you really do understand what the code is doing. Moreover make this clear to the reader.

4) Prove with code that you have indeed fixed the problem

By the end of finding and fixing all the errors you are to have a fully clean dataset/dataframe that you can write back to a file called Laurelhurst_crime_cleaned.csv. The ultimate goal, as it is for any data cleaning exercise, is to have high-quality data that can be used for analysis.  

In order to find the errors you should pay close attention to results of describe() for both the clean dataset and for the unclean dataset. Compare and contrast the results of each column. Also, spend some time looking at the format that each column is in and whether there are any outliers. You may also find it useful to apply the describe method to any subsets of erroneous data you get along the way. This is so you can understand your errors better.

We will provide some hints for each of the errors but the hints will become less precise as we reach the last error. 

You should insert all other code cells where appropriate.

### Error 1 (1.5 marks) ###
This error is concerned with an outlier in a numeric column. Look at what are valid values for each of the columns. See if you can find values that are not valid. Something went wrong with the data input here. Perhaps something to do with a decimal place? Identify that problem and then fix it. Insert appropriate code cells below.

### Error 2 (1.5 marks) ###
One thing we have not picked up previously - and which describe doesn't pick up - is whether we might have any issues with the datatypes. Firstly refresh your memory with the datatypes available for python and then check the datatypes in the unclean dataframe. There are three columns effected here and they all have the same problem. If we don't have them as the right datatype we won't be able to analyse our data properly. Insert appropriate code cells below.

### Error 3 (1.5 marks) ###
This error effects the Crime Against Category. Apply the unique() method against this column to see if all the values look right. Insert appropriate code cells below.

### Error 4 (2 marks) ###
This error effects two columns 'Offense Parent Group' and 'Offense'. Compare both the clean and the unclean datasets and the values that you see in these groups. Insert appropriate code cells below.

### Error 5 (3 marks) ###
This error effects the Report Number column. The hint for this is to look closely at the format for this column and try to identify anything that is not in this format. You should make any assumptions that you have about the format very clear. Insert appropriate code cells below. 